storage/src/main/java/io/delta/storage/GCSLogStore.java

```diff
@@ -0,0 +1,189 @@
+            Stream<StackTraceElement> extraStackTrace = Arrays.stream(
+                    realException.getStackTrace()).takeWhile(e ->
+                    !e.getClassName().contains(GCSLogStore.class.getSimpleName()));
+            // Combine the two stack traces, with a place holder just specifying that there
+            // was a helper method used, without any further details of the helper
+            Stream<StackTraceElement> placeHolderStackElem = Stream.of(
+                    new StackTraceElement(
+                            String.format(
+                                    "... run in separate thread using $s",
+                                    GCSLogStore.class.getSimpleName(),
+                                    " static method runInNewThread"),
+                            "",
+                            "",
+                            -1)
+            );
+            StackTraceElement[] finalStackTrace = Stream.concat(
+                    Stream.concat(baseStackTrace,
+                            placeHolderStackElem),
+                    baseStackTrace).toArray(StackTraceElement[]::new);
+            realException.setStackTrace(finalStackTrace);
+            throw realException;
+        } else {
+            return resultHolder.get(0);
+        }
+    }
+
+    @Override
+    public void write(Path path,
+                      Iterator<String> actions,
+                      Boolean overwrite,
+                      Configuration hadoopConf) throws IOException {
+        FileSystem fs = path.getFileSystem(hadoopConf);
+
+        // This is needed for the tests to throw error with local file system.
+        if (fs instanceof LocalFileSystem && !overwrite && fs.exists(path)) {
+            throw new FileAlreadyExistsException(path.toString());
+        }
+
+        // GCS may upload an incomplete file when the current thread is interrupted, hence we move
+        // the write to a new thread so that the write cannot be interrupted.
+        // TODO Remove this hack when the GCS Hadoop connector fixes the issue.
+        // If overwrite=false and path already exists, gcs-connector will throw
+        // org.apache.hadoop.fs.FileAlreadyExistsException after fs.create is invoked.
+        // This should be mapped to java.nio.file.FileAlreadyExistsException.
+        FSDataOutputStream stream = fs.create(path, overwrite);
+
+        Callable body = () -> {
+            try {
+                while (actions.hasNext()) {
+                    stream.write((actions.next() + "\n").getBytes(StandardCharsets.UTF_8));
+                }
+            } finally {
+                stream.close();
+
+            }
+            return "";
+        };
+
+        try {
+            runInNewThread("delta-gcs-logstore-write", true, body);
+        } catch (org.apache.hadoop.fs.FileAlreadyExistsException e) {
+            throw new FileAlreadyExistsException(path.toString());
```

Comment:
`.initCause(e)`

Reply:
removed FileAlreadyExistsException(..).initCause(e) that throws Throwable, to make it consistent with the other LogStore implementations